Feature/Added services page and resource API

dominiquekleeven opened the pull request:
Draft PR for #1686 and #1686

- Added registry mechanism for microservices/external services
- Added services page
- Added integration tests for the microservice resource and registry service

***
Branch still needs clean up (contains some docker changes for testing, which need to be removed/reset)


WIP

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/openremote/openremote?pullRequest=1979) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/openremote/openremote?pullRequest=1979) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/openremote/openremote?pullRequest=1979) it.</sub>

dominiquekleeven commented on manager/src/main/java/org/openremote/manager/microservices/MicroserviceResourceImpl.java:
```diff
@@ -0,0 +1,113 @@
+    public Response deregisterService(RequestParams requestParams, String serviceId, String instanceId) {
+        if (!isServiceAccount()) {
+            LOG.warning("Only service accounts can deregister services");
+            throw new ForbiddenException("Only service accounts can deregister services");
```
Changed

dominiquekleeven commented on manager/src/main/java/org/openremote/manager/microservices/MicroserviceResourceImpl.java:
```diff
@@ -0,0 +1,113 @@
+    public Response sendHeartbeat(RequestParams requestParams, String serviceId, String instanceId) {
+        if (!isServiceAccount()) {
+            LOG.warning("Only service accounts can send heartbeats");
+            throw new ForbiddenException("Only service accounts can send heartbeats");
```
Changed

dominiquekleeven commented:
Couple remaining issues regarding the front-end implementation:

- Refresh button placement in the header of the services page sidebar isn't properly centered and aligned
- Re-selecting a service does nothing, however this might be intended functionality to prevent accidental re-loading of the iframe.
- Unavailable services should not be able to be selected, or should provide some form of error message in the iframe area when selected
- The `or-iframe` component `onerror` event does not get triggered when the host is fully unreachable, e.g. URL doesn't lead anywhere, this causes the `or-iframe` component to get stuck in its loading state.

dominiquekleeven commented:
Deployed on `test3` with active service registration

dominiquekleeven commented on ui/component/or-services/src/or-service-tree.ts:
```diff
@@ -0,0 +1,144 @@
+          <or-mwc-input
+            type=${InputType.BUTTON}
+            icon="refresh"
+            title="Refresh services"
```
Resolved

dominiquekleeven commented on ui/app/manager/src/headers.ts:
```diff
@@ -37,6 +37,15 @@ export function headerItemInsights<S extends AppStateKeyed, A extends AnyAction>
+        icon: "puzzle",
+        href: "services",
+        text: "Services",
+        roles: () => !manager.hasRealmRole("restricted_user")
```
Resolved

dominiquekleeven commented on manager/src/main/java/org/openremote/manager/microservices/MicroserviceRegistryService.java:
```diff
@@ -0,0 +1,272 @@
+
+    protected void cleanupExpiredUnavailableInstances() {
+        long currentTime = timerService.getCurrentTimeMillis();
+        long purgeThreshold = currentTime + DEFAULT_PURGE_UNAVAILABLE_MS;
```
Yes, indeed a bug. 

Logic here:
- get current timestamp
- subtract the purge threshold e.g. 24 hours
- iterate through the services and collect any that are unavailable and have an expiration timestamp that is older than the purge timestamp.
- de-register the collected services

dominiquekleeven commented:
Will update `test3` tomorrow with the latest changes, cleaning up the PR and removing any unrelated files, and changing to status to Ready for review

dominiquekleeven commented:
Deployed on `test3` alongside the ML Forecast Service (https://github.com/openremote/service-ml-forecast/pull/40 version)

MartinaeyNL commented on ui/component/or-services/src/index.ts:
Can you use an indentation of 4 n `.ts` files, like the other components?
(an automatic linter is being worked on 😂 )

dominiquekleeven replied:
They all should be indented using 4 spaces now, I updated the preffier config we had in the repo to enforce so for those who use prettier it'll automatically enforce 4 spaces for identation when formatting.

MartinaeyNL commented on ui/component/or-services/src/index.ts:
I was testing the UI, and i tend to press the "reload services" button, just "to double check if everything is loaded".
When everything is OK, there is nothing happening in the UI.

Maybe we should show a simple snackbar saying; "Loaded X services!", to indicate that 'something' is going on?

dominiquekleeven replied:
Good idea, I've added a snackbar for this and the translations

It's now `{{count}} service(s) retrieved`  / `{{count}} service(s) opgehaald`

MartinaeyNL reviewed (commented):
I'm overall very happy with the code 👍 
Mostly went over the frontend related code, but also had a quick look at the other code.
Let me know if you want to go over stuff in a call.

I also think some slight changes in CSS could improve the looks of this page.
But I think you have a chat scheduled on that next week, I believe?

dominiquekleeven commented on ui/component/or-components/src/or-iframe.ts:
```diff
@@ -0,0 +1,217 @@
+  public timeout = 10000; // 10 seconds default timeout
+
+  @property({ type: String })
+  public loadErrorMessage?: string;
```
Replaced it with a slot element :)

dominiquekleeven commented on ui/component/or-components/src/or-iframe.ts:
```diff
@@ -0,0 +1,217 @@
+  private loading = true;
+
+  @state()
+  private loadingError = false;
```
Changed to just error

dominiquekleeven commented on ui/component/or-components/src/or-iframe.ts:
```diff
@@ -0,0 +1,217 @@
+      case OrIFrameEventType.LOADED:
+        this.loadingError = false;
+        this.loading = false;
+        this.requestUpdate();
```
Removed it

dominiquekleeven commented on model/src/main/java/org/openremote/model/microservices/MicroserviceResource.java:
```diff
@@ -0,0 +1,118 @@
+    @Consumes(APPLICATION_JSON)
+    @Produces(APPLICATION_JSON)
+    @RolesAllowed({ Constants.WRITE_SERVICES_ROLE })
+    @Operation(operationId = "registerService", summary = "Register an external service/microservice", responses = {
+            @ApiResponse(responseCode = "200", description = "Service registered successfully", content = @Content(mediaType = "application/json", schema = @Schema(implementation = Microservice.class))),
+            @ApiResponse(responseCode = "400", description = "Invalid microservice object"),
+            @ApiResponse(responseCode = "409", description = "Microservice instance already registered"),
+    })
```
Yea I think I saw these being used in another resource interface so basically used that as inspiration :)

dominiquekleeven commented on ui/component/or-services/src/or-service-tree.ts:
```diff
@@ -0,0 +1,164 @@
+  }
+
+  protected _onRefreshServices(): void {
+    this.dispatchEvent(new CustomEvent("refresh-services", { detail: {} }));
```
Added interfaces for all the new CustomEvent I've been using, also added the @event decorator to the components

dominiquekleeven commented:
@MartinaeyNL Thanks for the review!

I've addressed your comments and `test3` has been updated with the changes, feel free to review again when you have the time.

dominiquekleeven commented:
`test3` updated with latest changes

MartinaeyNL commented on ui/app/manager/src/pages/page-services.ts:
```diff
@@ -0,0 +1,344 @@
+ */
+import { css, html, TemplateResult, PropertyValues } from "lit";
+import { customElement, state } from "lit/decorators.js";
+import { Page, PageProvider, router } from "@openremote/or-app";
+import { AppStateKeyed } from "@openremote/or-app";
```
You can merge these 😉

MartinaeyNL commented on ui/app/manager/src/pages/page-services.ts:
```diff
@@ -0,0 +1,344 @@
+        }
+
+        try {
+            const realmServices = await manager.rest.api.MicroserviceResource.getServices({ realm: this.realmName });
+            const globalServices = await manager.rest.api.MicroserviceResource.getGlobalServices();
```
I'd prefer to use something like `Promise.all()` which executes these HTTP requests in parallel.
On another note; if the global request fails, it "fails to load services", although realm services were fetched correctly.
Would be nice to do some minor tweaks here.

dominiquekleeven replied:
I opted for `Promise.allSettled` according to docs it allows it to handle the individual promise failures, so it now runs the requests in parallel and handles the individual failures accordingly. It is very unlikely that one succeeds and the other fails, but it'll handle it accordingly. Also handles the case where both fail

MartinaeyNL commented on setup/src/integration/java/org/openremote/setup/integration/KeycloakTestSetup.java:
```diff
@@ -83,6 +85,8 @@ public KeycloakTestSetup(Container container) {
     public void onStart() throws Exception {
         super.onStart();
 
+        ManagerIdentityProvider identityProvider = identityService.getIdentityProvider();
+
         // Realms
         realmMaster = identityService.getIdentityProvider().getRealm(Constants.MASTER_REALM);
```
Seems like a duplicate pattern; we should between
- Caching `identityProvider` in an variable, or..
- use `identityService.getIdentityProvider()` throughout the setup code.

MartinaeyNL reviewed (commented):
Reviewed the UI code, and took a short look at the backend code.
Only had a few minor comments. Great stuff! 👍

MartinaeyNL reviewed (commented):
The comments I had have been resolved. So I'm OK with these changes 👍 
Let's have Rich take a deeper look into the manager code and the overall solution.

dominiquekleeven commented on ui/.prettierrc.json:
```diff
@@ -1,5 +1,7 @@
   "trailingComma": "es5",
   "printWidth": 120,
-  "singleQuote": false
+  "singleQuote": false,
+  "tabWidth": 4,
+  "useTabs": false
```
Should be fine, we don't apply formatting yet and 4 spaces is what we use for indentation.

dominiquekleeven commented on manager/src/main/java/org/openremote/manager/microservices/MicroserviceRegistryService.java:
```diff
@@ -0,0 +1,383 @@
+        LOG.info("Successfully registered microservice: " + microservice.getServiceId() + ", instanceId: "
+                + microservice.getInstanceId() + ", isGlobal: " + microservice.getIsGlobal());
+
+        return microservice;
```
Yep totally missed that, been dabbling too much in programming languages that pass by value rather than instance.

Fixed it :)

dominiquekleeven commented on manager/src/main/java/org/openremote/manager/microservices/MicroserviceResourceImpl.java:
```diff
@@ -0,0 +1,214 @@
+        try {
+            Microservice registeredMicroservice = microserviceRegistry.registerService(microservice);
+
+            if (registeredMicroservice == null) {
```
Removed the redundant null check

dominiquekleeven commented on manager/src/main/java/org/openremote/manager/microservices/MicroserviceResourceImpl.java:
```diff
@@ -0,0 +1,214 @@
+        try {
+            Microservice registeredMicroservice = microserviceRegistry.registerService(microservice, true);
+
+            if (registeredMicroservice == null) {
```
Removed the redundant null check

dominiquekleeven commented on manager/src/main/java/org/openremote/manager/microservices/MicroserviceRegistryService.java:
```diff
@@ -0,0 +1,383 @@
+
+        if (microservice.getInstanceId() == null) {
+            microservice.setInstanceId(UniqueIdentifierGenerator.generateId());
+        }
```
Moved before logging

dominiquekleeven commented on manager/src/main/java/org/openremote/manager/microservices/MicroserviceRegistryService.java:
```diff
@@ -0,0 +1,383 @@
+     * @return The microservice
+     */
+    public Microservice getService(String serviceId, String instanceId) {
+        return registry.values().stream()
```
Made it a direct get

dominiquekleeven commented on manager/src/main/java/org/openremote/manager/microservices/MicroserviceRegistryService.java:
```diff
@@ -0,0 +1,383 @@
+
+    // serviceId: <instanceId: Microservice>
+    // Nested ConcurrentHashMaps for segmenting, segment locks and fast lookups
+    protected ConcurrentHashMap<String, ConcurrentHashMap<String, Microservice>> registry;
```
You’re right, the nested concurrent map was unnecessary for synchronization. I’ve changed it to a regular map and made sure that any lookups using the instanceId are done on the key.
